src/Strategy/Python/Poetry.hs

```diff
@@ -0,0 +1,187 @@
+    , projectLicenses = pure []
+    }
+
+getDeps :: (Has ReadFS sig m, Has Diagnostics sig m, Has Logger sig m) => PoetryProject -> m (Graphing Dependency)
+getDeps project = do
+  context "Poetry" $ context "Static analysis" $ analyze project
+
+-- | Analyzes Poetry Project and creates dependency graph.
+analyze ::
+  ( Has ReadFS sig m
+  , Has Diagnostics sig m
+  , Has Logger sig m
+  ) =>
+  PoetryProject ->
+  m (Graphing Dependency)
+analyze PoetryProject{pyProjectToml, poetryLock} = do
+  pyproject <- readContentsToml pyProjectCodec (pyProjectTomlPath pyProjectToml)
+  case poetryLock of
+    Just lockPath -> do
+      poetryLockProject <- readContentsToml poetryLockCodec (poetryLockPath lockPath)
+      _ <- logIgnoredDeps pyproject (Just poetryLockProject)
+      context "Building dependency graph from pyproject.toml and poetry.lock" $ pure $ buildGraphWithLock poetryLockProject pyproject
+    Nothing -> context "Building dependency graph from only pyproject.toml" $ pure $ Graphing.fromList $ pyProjectDeps pyproject
+
+-- | Build graph of `Dependency` only from `PyProject` and `PoetryLock`.
+buildGraphWithLock :: PoetryLock -> PyProject -> Graphing Dependency
+buildGraphWithLock lockProject poetryProject = promoteToDirect isDirect graph
+  where
+    -- Dependencies in `poetry.lock` are direct if they're specified in `pyproject.toml`.
+    isDirect :: Dependency -> Bool
+    isDirect dep = case pyprojectPoetry poetryProject of
+      Nothing -> False
+      Just _ -> any (\n -> dependencyName n == dependencyName dep) (pyProjectDeps poetryProject)
+
+    lowerCasedPkgName :: PackageName -> PackageName
+    lowerCasedPkgName name = PackageName . toLower $ unPackageName name
+
+    graph = gmap pkgNameToDependency (buildPackageNameGraph $ poetryLockPackages lockProject)
+    mapOfDependency = toMap $ poetryLockPackages lockProject
+
+    -- Pip packages are [case insensitive](https://www.python.org/dev/peps/pep-0508/#id21), but poetry.lock may use
+    -- non-canonical name for reference. Try to lookup with provided casing, otherwise fallback to lower casing.
+    --
+    -- Poetry.lock when referencing deep dependencies uses non canonical names.
+    --
+    --  ```toml
+    --  [package.dependencies]
+    --  MarkupSafe = ">=2.0"
+    --  ....
+    --
+    -- [[package]]
+    -- name = "markupsafe"
+    -- version = "2.0.1"
+    -- ...
+    -- ```
+    --
+    -- TODO: Better approach to handle casing scenarios with poetry.lock file.
+    pkgNameToDependency name =
+      fromMaybe
+        ( Dependency
+            { dependencyType = PipType
+            , dependencyName = unPackageName name
+            , dependencyVersion = Nothing
+            , dependencyLocations = []
+            , dependencyEnvironments = []
+            , dependencyTags = Map.empty
+            }
+        )
+        $ Map.lookup name mapOfDependency
+          <|> Map.lookup (lowerCasedPkgName name) mapOfDependency
```

Comment:
Given your stated intentions, I would move this into `buildPackageNameGraph` such that `buildPackageNameGraph :: [PoetryLockPackage] -> Graphing Dependency` because:

1. This logic is only based on lockfile information, and doesn't require the pyproject.toml, so it's clearer that the intention of `buildPackageNameGraph` is "construct a graph from `poetry.lock`'s information".
2. This makes it clearer that `buildGraphWithLock` is just for setting direct dependencies using `pyproject.toml`.

I might also:

1. Rename `buildPackageNameGraph` to `graphFromLockfile :: [PoetryLockPackage] -> Graphing Dependency`.
2. Rename `buildGraphWithLock` to `setGraphDirectsFromPyproject :: Graphing Dependency -> PyProject -> Graphing Dependency`.
3. Call these two functions in `analyze` as appropriate.